File: Ch01/ex1_10.erl
-module(ex1_10).
-author("Kevin C. Baird").
-purpose("SICP in Erlang: Ackermann Function").
-compile(export_all).

ackermann(_, 0) -> 0;
ackermann(0, Y) -> Y * 2;
ackermann(_, 1) -> 2;
ackermann(X, Y) -> ackermann(X-1, ackermann(X, Y-1)).

f(N) -> ackermann(0, N). % x * 2
g(N) -> ackermann(1, N). % 2 ** x if x > 0; 0 if x == 0; error if x < 0
h(N) -> ackermann(2, N). % [0 2 4 16 65536] read at index x if 0 <= x <= 4; error otherwise
k(N) -> 5 * N * N.       % 5 n squared

File: Ch01/Ex1_10.hs
{-
 - Ex1_10.hs
 - Kevin C. Baird
 - SICP in Haskell
 -
 - Ackermann function
 - (Exercise 1.10)
 -
 -}

module Ex1_10 where

ackermann x y
  | y == 0 = 0
  | x == 0 = y * 2
  | y == 1 = 2
  | otherwise = ackermann (x-1) (ackermann x (y-1))

f   = ackermann 0 -- x * 2
g   = ackermann 1 -- 2 ** x if x > 0; 0 if x == 0; error if x < 0
h   = ackermann 2 -- [0 2 4 16 65536] read at index x if 0 <= x <= 4; error otherwise
k n = 5 * n * n     -- 5 n squared

